pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -528,21 +540,31 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 				set.Namespace,
 				set.Name,
 				replicas[target].Name)
-			err := ssc.podControl.DeleteStatefulPod(set, replicas[target])
+			if err := ssc.podControl.DeleteStatefulPod(set, replicas[target]); err != nil {
+				return &status, err
+			}
+			// After deleting a Pod, dont Return From Here Yet.
+			// We might have maxUnavailable greater than 1
 			status.CurrentReplicas--
-			return &status, err
 		}
 
-		// wait for unhealthy Pods on update
-		if !isHealthy(replicas[target]) {
+		if getPodRevision(replicas[target]) != updateRevision.Name || !isHealthy(replicas[target]) {
```

Comment:
Are old but healthy pods counted against maxUnavailable?

Reply:
maxUnavailable will always be unhealthy pods. If they are healthy and on old revision, its not counted against maxUnavailable

Reply:
@lavalamp sorry my bad, the Deleted pod above will get added to unavailablePods by this check. Specifically !IsHealthy() will catch it which has an internal check for terminating pod and Delete will mark a pod as terminating,.

Reply:
That doesn't seem plausible since isHealthy is looking at a cached pod and not talking to apiserver again.

Reply:
The previous code didn't check the revision. Why are we checking now?